### ISSUE TITLE
Add types definition

### ISSUE THREAD

**Issue by VitorLuizC** [VitorLuizC/uncouple#2]
Add types definition.

Wait until TypeScript 2.8 to use conditional types.

```ts
declare module 'uncouple' {
  type Return<F> = F extends ((...args: (infer A)[]) => infer R) ? R : never

  type Uncouple<T extends Object, F> = (instance: T, ...args: any[]) => Return<F>

  type Methods<T extends Object> = {
    -readonly [P in keyof T]: T[P] extends (...args: any[]) => Return<T[P]>
      ? Uncouple<T, T[P]>
      : undefined;
  }

  function uncouple <T extends Object> (object: T): Methods<T>

  export default uncouple
}
```

**Comment by VitorLuizC** [comment]
I don't know yet how to dynamically generate arguments.

I tried same strategy as Return and it doesn't work : /

```ts
  type Args<F> = F extends ((a1: infer A1) => any)                             ? [A1]
               : F extends ((a1: infer A1, a2: infer A2) => any)               ? [A1, A2]
               : F extends ((a1: infer A1, a2: infer A2, a3: infer A3) => any) ? [A1, A2, A3]
               : F extends ((...args: (infer A)[]) => any)                     ? A
                                                                               : any

type Uncouple<T extends Object, F> = (instance: T, args: Args<F>) => Return<F>
```

**Comment by VitorLuizC** [comment]
I probably find it here.

https://github.com/Microsoft/TypeScript/issues/5453

**Comment by VitorLuizC** [comment]
Arguments will be implemented in TypeScript 3.0 https://github.com/Microsoft/TypeScript/pull/24897

Until this I can just type return and method name
```ts
declare type Method = (...args: any[]) => any;

declare type Uncouple <T extends Object, U extends Method> = (instance: T, ...args: any[]) => ReturnType<U>;

declare type UncoupledMethods <T extends Object> =
  { [K in keyof T]: T[K] extends Method ? Uncouple<T, T[K]> : never };

declare function uncouple <T extends Object> (object: T): UncoupledMethods<T>;
```
